Fix segfaults with string_replace and string_ireplace

marcioAlmada opened the pull request:
Like in `ext/standard/tests/strings/str_replace_array_refs.phpt`:
```php
<?php
$data = ['a' => 'b'];
$ref = &$data;
$b = &$ref['a'];
var_dump(str_replace(array_keys($data), $data, "a"));
// segfault :~
```

And other possible very obscure situations involving object properties (with array values) being passed to `str_replace`:

```php
<?php
$closure = function (array $array, array $keys, $value) {
    $current = &$array;
    foreach ($keys as $key) $current = &$current[$key];
    $current = $value;
    return $array;
};
class SomeClass { public $prop; }
$obj = new SomeClass;
$obj->prop = ['x' => 'property'];
$obj->prop = $closure($obj->prop, ['x'], 'a');
var_dump(str_replace(array_keys($obj->prop), $obj->prop, "x property"));
```

PS: Problem is not present on PHP-5 and before, bug is only on master AFAIK.

marcioAlmada commented:
Apart from the segfault correction:

PR deals with behavior related to references that was not specified anywhere on documentation. So the BC compatible route was taken: test is now more strict and fully compatible with PHP-5.6 and previous outputs, in case any user land code is relying on it.

laruence commented on ext/standard/string.c:
```diff
@@ -4006,7 +4007,9 @@ static zend_long php_str_replace_in_subject(zval *search, zval *replace, zval *s
 				if ((replace_entry = zend_hash_get_current_data_ex(Z_ARRVAL_P(replace), &pos)) != NULL) {
 					/* Make sure we're dealing with strings. */
-					convert_to_string_ex(replace_entry);
+					ZVAL_DEREF(replace_entry);  // Deal with possibly shared zvals
```
replace_entry is readonly, maybe use zval_get_string here?

marcioAlmada replied:
zval_get_string segfaults if done before [L402](https://github.com/php/php-src/pull/1022/files#diff-b31234a9f5a03a328b60d0453988140fR4012) (convert_to_string), don't know for sure why.

I'll try that again, anyway.

marcioAlmada replied:
Just noticed that `zval_get_string` won't cause the same behavior from PHP-5.6, and backwards, since `string_replace` wouldn't cast each array val (from $search and $replaces args) to string anymore.

Should we change that collateral behavior?

laruence replied:
we should use the more efficient one. 

anyway,  seems you test script doesn't covert this case "string to array casting?"

marcioAlmada replied:
The array to string conversion case is already covered by the other tests and behavior is not affected by this PR.

marcioAlmada commented:
Any news about https://github.com/php/php-src/pull/1022#discussion_r23493863 ?

laruence commented on ext/standard/string.c:
```diff
@@ -3990,7 +3990,8 @@ static zend_long php_str_replace_in_subject(zval *search, zval *replace, zval *s
 		ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(search), search_entry) {
 			/* Make sure we're dealing with strings. */
-			SEPARATE_ZVAL(search_entry);
+			ZVAL_DEREF(search_entry); // Deal with possibly shared zvals
```
minor one, please don't use "//" as comment, use "/*". or maybe remove this moment at all..

marcioAlmada commented:
@laruence done:

- removed cpp comments
- most efficient solution used + we have a more sane behavior.
- tests updated

Beware branch was cleaned up and rebased with master so a forced push was required.

laruence commented:
merged, thanks
